hathor/indexes/memory_tx_group_index.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2021 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from collections import defaultdict
+from typing import DefaultDict, Iterable, Set, TypeVar
+
+from structlog import get_logger
+
+from hathor.indexes.tx_group_index import TxGroupIndex
+from hathor.transaction import BaseTransaction
+
+logger = get_logger()
+
+KT = TypeVar('KT')
+
+
+class MemoryTxGroupIndex(TxGroupIndex[KT]):
+    """ Index of inputs/outputs by address
+    """
+
+    index: DefaultDict[KT, Set[bytes]]
+
+    def __init__(self) -> None:
+        self.force_clear()
+
+    def force_clear(self) -> None:
+        self.index = defaultdict(set)
+
+    def _add_tx(self, key: KT, tx: BaseTransaction):
+        self.index[key].add(tx.hash)
+
+    @abstractmethod
+    def _extract_keys(tx: BaseTransaction) -> Iterable[KT]:
+        raise NotImplementedError
+
+    def add_tx(self, tx: BaseTransaction) -> None:
+        """ Add tx inputs and outputs to the wallet index (indexed by its addresses).
+        """
+        assert tx.hash is not None
+
+        for key in self._extract_keys(tx):
+            self._add_tx(key, tx)
+
+    def remove_tx(self, tx: BaseTransaction) -> None:
+        """ Remove tx inputs and outputs from the wallet index (indexed by its addresses).
+        """
+        assert tx.hash is not None
+
+        addresses = tx.get_related_addresses()
+        for address in addresses:
+            self.index[address].discard(tx.hash)
```

Comment:
This implementation doesn't seem like it's generic over the key type, since it's using "address" which is `str`.

Reply:
Fixed! Thanks!